proposals/csharp-8.0/async-streams.md

```diff
@@ -187,52 +187,78 @@ await foreach (var i in enumerable)
+  - Member lookup is performed on `E` with the identifier `MoveNextAsync` and no type arguments. If the member lookup produces no match, the result is an error, or the result is anything except a method group, an error is produced and no further steps are taken.
+  - Overload resolution is performed on the method group with an empty argument list. If overload resolution results in no applicable methods, results in an ambiguity, or results in a single best method but that method is either static or not public, or its return type is not awaitable into `bool`, an error is produced and no further steps are taken.
+  - The collection type is `X`, the enumerator type is `E`, and the iteration type is the type of the `Current` property.
+- Otherwise, check for an enumerable interface:
+  - If among all the types `Tᵢ` for which there is an implicit conversion from `X` to `IAsyncEnumerable<ᵢ>`, there is a unique type `T` such that `T` is not dynamic and for all the other `Tᵢ` there is an implicit conversion from `IAsyncEnumerable<T>` to `IAsyncEnumerable<Tᵢ>`, then the collection type is the interface `IAsyncEnumerable<T>`, the enumerator type is the interface `IAsyncEnumerator<T>`, and the iteration type is `T`.
+  - Otherwise, if there is more than one such type `T`, then an error is produced and no further steps are taken.
+- Otherwise, an error is produced and no further steps are taken.
 
-This code:
+The above steps, if successful, unambiguously produce a collection type `C`, enumerator type `E` and iteration type `T`.
 
 ```csharp
-var enumerable = ...;
-await foreach (T item in enumerable)
-{
-   ...
-}
+await foreach (V v in x) «embedded_statement»
 ```
 
-is translated to the equivalent of:
+is then expanded to:
 
 ```csharp
-var enumerable = ...;
-var enumerator = enumerable.GetAsyncEnumerator();
-try
 {
-    while (await enumerator.MoveNextAsync())
-    {
-       T item = enumerator.Current;
-       ...
+    E e = ((C)(x)).GetAsyncEnumerator();
+    try {
+        while (await e.MoveNextAsync()) {
+            V v = (V)(T)e.Current;
+            «embedded_statement»
+        }
+    }
+    finally {
+        ... // Dispose e
     }
-}
-finally
-{
-    await enumerator.DisposeAsync(); // omitted, along with the try/finally, if the enumerator doesn't expose DisposeAsync
 }
 ```
 
-If the iterated type doesn't expose the right pattern, the interfaces will be used.
+The body of the `finally` block is constructed according to the following steps:
+- If the type `E` has an appropriate `DisposeAsync` method:
+  - Perform member lookup on the type `E` with identifier `DisposeAsync` and no type arguments. If the member lookup does not produce a match, or it produces an ambiguity, or produces a match that is not a method group, check for the disposal  interface as described below.
```

Comment:
>disposal  interface [](http://example.com/codeflow?start=234&length=19)

Extra space between words. And should this be "disposable interface" (here and below) instead?

Reply:
I'm trying to go for parallel with "enumeration interfaces".
"disposal" and "enumeration" are nouns, but "disposable" and "enumerable" are adjectives?